### DISCUSSION TITLE
Re-render issue while performing open and setOpen using shadcn

### DISCUSSION THREAD

**Discussion by Akshaypmna18** [pmndrs/zustand#2354] [Bug report] [answered]
https://stackoverflow.com/questions/77826561/shadcn-reusuable-dialog-component-called-twice-renders-update-form-only

I have implemented a reusable dialog component using **Shadcn** in my todo app. However, I'm encountering an issue where the component is not working as needed. The add form and update form opens only after clicking twice. And when we click the pencil icon immediately after adding a todo it will pop add form instead of update form. I'm using **zustand** for state management.

Dialog Component

``` 
export default function DialogModal({ children, id, text }) {
  const { handleTodo, todo, setTodo, isOpen, setIsOpen } = useTodo(
    (state) => state
  );
  return (
    <Dialog open={isOpen} onOpenChange={() => setIsOpen()}>
      <DialogTrigger>{children}</DialogTrigger>
      <DialogContent className="max-w-[400px]">
        <DialogHeader>
          <DialogTitle className="bold text-[calc(1.25rem+.5vw)]">
            <big>{id ? "Update Task" : "Add Task"}</big>
          </DialogTitle>
        </DialogHeader>
        <Input
          className="font-[poppins]"
          placeholder="Enter the task..."
          defaultValue={text}
          onChange={(e) => setTodo(e.target.value)}
          onKeyDown={(e) => {
            if (e.key === "Enter") handleTodo(todo.trim(), id);
          }}
        />
        <DialogFooter>
          <Button
            className="text-[calc(1rem+.5vw)] mx-auto w-[min(90%,10rem)]"
            onClick={() => handleTodo(todo.trim(), id)}
          >
            {id ? "Update" : "Add"}
          </Button>
        </DialogFooter>
      </DialogContent>
    </Dialog>
  );
}
```
If I set 'setIsOpen(!isOpen)' the dialog component only shows the latest todo

update todo

```
     <DialogModal id={id} text={text}>
      <ToolTipComp Content={() => <ToolTipContent />}>
        <Pencil2Icon
          onClick={() => setIsOpen(true)}
          className="min-h-[1.5rem] min-w-[1.5rem] cursor-pointer hover:text-rose-500"
        />
      </ToolTipComp>
    </DialogModal>
```

add todo

```
    <DialogModal>
      <Button
        onClick={() => setIsOpen(true)}
        className="font-[poppins] fixed right-[calc(2.5rem+1vw)] bottom-[calc(3rem+1vh)] text-[calc(2rem+1vw)] rounded-full h-[calc(2.5rem+1vw)] w-[calc(2.5rem+1vw)]"
      >
        +
      </Button>
    </DialogModal>
```

[demo](https://aks-to-do.vercel.app/)

[dialog component code](https://github.com/Akshaypmna18/to-do/blob/main/src/components/dialogModal/DialogModal.jsx)

[add todo code](https://github.com/Akshaypmna18/to-do/blob/main/src/features/todos/addTodo/AddTodo.jsx)

[update todo code](https://github.com/Akshaypmna18/to-do/blob/main/src/features/todos/list/TodosEle.jsx)

[zustand store code](https://github.com/Akshaypmna18/to-do/blob/main/src/store.jsx)

[video for better understanding](https://drive.google.com/file/d/1KWMQ76oWjI_DZ4gjLrlbLDV7_dzKA2wb/view?usp=drive_link)

Use devices with a width below 640 pixels or adjust it in the browser.

**Comment by dbritto-dev** [comment]
@Akshaypmna18 let me look into that, I've got exp with Radix Ui and Shadcn UI

**Reply by Akshaypmna18** [reply]
Thanks pal.

**Reply by dbritto-dev** [reply]
@Akshaypmna18 here you go -> https://github.com/Akshaypmna18/to-do/pull/1

BTW, there's no issues with zustand the issue was related to a bad implementation of radix ui or shadcn ui

Few things:
- If you want to create components on top of radix ui or shadcn ui, you need to keep in mind that you should respect the composition abstraction, if there is a over abstraction of the components you could brake that composition abstraction and you could face some issues or would be difficult to compose some components on top of each others
- You should always use the next state on change events when you have controlled components, for instance: `isOpen` and `onOpenChange`

Here's an implementation of shadcn ui: tooltip + dialog but can be tooltip + popover  or tooltip + alert dialog

```jsx
import { forwardRef } from "react";
import * as TooltipShadcn from "../ui/tooltip";
import * as DialogShadcn from "../ui/dialog";
import * as InputShadcn from "../ui/input";
import * as ButtonShadcn from "../ui/button";
import * as LabelShadcn from "../ui/label";

export const Tooltip = forwardRef((props, ref) => {
  return (
    <TooltipShadcn.TooltipProvider>
      <TooltipShadcn.Tooltip ref={ref} {...props} />
    </TooltipShadcn.TooltipProvider>
  );
});

export const TooltipTrigger = forwardRef((props, ref) => {
  return <TooltipShadcn.TooltipTrigger ref={ref} {...props} asChild />;
});

export const TooltipContent = forwardRef((props, ref) => {
  return <TooltipShadcn.TooltipContent ref={ref} {...props} />;
});

export function TooltipDialogDemo() {
  return (
    <Tooltip>
      <DialogShadcn.Dialog>
        <DialogShadcn.DialogTrigger asChild>
          <TooltipTrigger>
            <ButtonShadcn.Button variant="outline">
              Edit Profile
            </ButtonShadcn.Button>
          </TooltipTrigger>
        </DialogShadcn.DialogTrigger>
        <DialogShadcn.DialogContent className="sm:max-w-[425px]">
          <DialogShadcn.DialogHeader>
            <DialogShadcn.DialogTitle>Edit profile</DialogShadcn.DialogTitle>
            <DialogShadcn.DialogDescription>
              Make changes to your profile here. Click save when you're done.
            </DialogShadcn.DialogDescription>
          </DialogShadcn.DialogHeader>
          <div className="grid gap-4 py-4">
            <div className="grid grid-cols-4 items-center gap-4">
              <LabelShadcn.Label htmlFor="name" className="text-right">
                Name
              </LabelShadcn.Label>
              <InputShadcn.Input
                id="name"
                value="Pedro Duarte"
                className="col-span-3"
              />
            </div>
            <div className="grid grid-cols-4 items-center gap-4">
              <LabelShadcn.Label htmlFor="username" className="text-right">
                Username
              </LabelShadcn.Label>
              <InputShadcn.Input
                id="username"
                value="@peduarte"
                className="col-span-3"
              />
            </div>
          </div>
          <DialogShadcn.DialogFooter>
            <ButtonShadcn.Button type="submit">
              Save changes
            </ButtonShadcn.Button>
          </DialogShadcn.DialogFooter>
        </DialogShadcn.DialogContent>
      </DialogShadcn.Dialog>
      <TooltipContent>Just a random tooltip</TooltipContent>
    </Tooltip>
  );
}
```

**Reply by Akshaypmna18** [reply]
Thank you for your insights on Radix UI and Shadcn UI. I'll ensure to respect composition abstraction and use the next state on change events, following your valuable advice. Thanks for your expertise and feedback. Open to more insights!

**Reply by Akshaypmna18** [reply]
@dbritto-dev I have gone through your code. And it is sad to say that the code is not working as intended.

[video](https://drive.google.com/file/d/1pjYqV1RK2ci8soWlLOQ-QuxJlb3euQHf/view?usp=sharing)

**Reply by dbritto-dev** [reply]
> @dbritto-dev I have gone through your code. And it is sad to say that the code is not working as intended.
> 
> 
> 
> [video](https://drive.google.com/file/d/1pjYqV1RK2ci8soWlLOQ-QuxJlb3euQHf/view?usp=sharing)

What's the issue? Looks good to me. The previous version wasn't working because you need to do double click the first time

**Reply by dbritto-dev** [reply, marked as answer]
@Akshaypmna18 here you go -> https://github.com/Akshaypmna18/to-do/pull/2 (the fix for the issue related to dialog showing up the latest todo)

**Reply by Akshaypmna18** [reply]
Great pal

**Comment by dai-shi** [comment]
https://github.com/pmndrs/zustand/issues/2356#issuecomment-1961029029

**Reply by dbritto-dev** [reply]
@dai-shi nothing to do with zustand, the issue was related to a poor implementation -> https://github.com/Akshaypmna18/to-do/pull/2 I could explain what was the issue and the fixes if is necessary

**Comment by Akshaypmna18** [comment]
@dai-shi sorry for the trouble. I thought it was the issue of zustand
